lib/matplotlib/dates.py

```diff
@@ -740,12 +746,232 @@ def __call__(self, x, pos=0):
+        ticks = self._locator()
+        if pos is not None:
+            if not np.array_equal(ticks, self._oldticks):
+
+                offset_fmt = ''
+                fmt = self.defaultfmt
+                self._formatter = DateFormatter(fmt, self._tz)
+                tickdate = np.array([num2date(tick).timetuple()[:6]
+                                     for tick in ticks])
+                tickdatetime = [num2date(tick) for tick in ticks]
+
+                # basic algorithm:
+                # 1) only display a part of the date if it changes over the
+                #    ticks.
+                # 2) don't display the smaller part of the date if:
+                #    it is always the same or if it is the start of the
+                #    year, month, day etc.
+                level = 5
+                diffed = False
+                # fmt for most ticks at this level
+                fmts = self.formats
+                # format beginnings of days, months, years, etc...
+                zerofmts = self.zero_formats
+                # offset fmt are for the offset in the upper left of the
+                # or lower right of the axis.
+                offsetfmts = self.offset_formats
+
+                # determine the level we will label at:
+                # mostly 0: years,  1: months,  2: days,
+                # 3: hours, 4: minutes, 5: seconds, 6: microseconds
+                while not diffed and level >= 0:
+                    if np.any(tickdate[:, level][:-1] !=
+                              tickdate[:, level][1:]):
+                        diffed = True
+                    if not diffed:
+                        level -= 1
+
+                # level is the basic level we will label at.
+                # now loop through and decide the actual ticklabels
+                zerovals = [0, 1, 1, 0, 0, 0, 0]
+                ticknew = []
+                zeroed = False
+                for nn in range(len(tickdate)):
+                    if tickdate[nn][level] == zerovals[level]:
+                        fmt = zerofmts[level]
+                        zeroed = True
+                    else:
+                        fmt = fmts[level]
+                    newtickla = tickdatetime[nn].strftime(fmt)
+                    ticknew.append(newtickla)
+
+                # special handling of seconds and microseconds:
+                # strip extra zeros and decimal if possible...
+                if level >= 5:
+                    for ind in reversed(range(2, 9)):
+                        if any(len(tn) == 9 and
+                                tn[ind] != '0' for tn in ticknew):
+                                break
+                    if ind == 2:
+                        # strip the decimal as well if we got to here.
+                        ind = 1
+                    for nn in range(len(ticknew)):
+                        if len(ticknew[nn]) == 9:
+                            ticknew[nn] = ticknew[nn][:ind+1]
```

Comment:
Just replace the whole thing by
```
if level >= 5:
    for nn in range(len(ticknew)):
        ticknew[nn] = ticknew[nn].rstrip("0").rstrip(".")
```
?
(it's not totally clear to me where the 9 comes from, in fact).
(and yes you can even write `.rstrip(".0")` which will remove all dots and zeros but I always find this a bit confusing)

Reply:
It wasn't *quite* that simple, but the new version should be better (you can't strip out any trailing zeros before the decimal place)

Reply:
Goops, gotta go, but realized this isn't quite what I want either - I want all numbers to have the same decimal place precision, even if all zeros.. So this needs a bit of modification to be right...

Reply:
> (you can't strip out any trailing zeros before the decimal place)

OK, so rstrip zeros first, then dots later? (the version I gave above, basically)

(Not taking into account your second comment...)

Reply:
ie if I need `1.5`, then I should also have `1.0` and `2.0`

Reply:
> > (you can't strip out any trailing zeros before the decimal place)
> 
> OK, so rstrip zeros first, then dots later? (the version I gave above, basically)

Second comment aside, even at level 5, I have ticks that are mixed in format, ie. for seconds around 3 am:  `03:00`, `01.0000000`, `02.000000`, so I need to check for a dot first or the 3am string would be `03:`.

Reply:
Ah yes, I completely missed that.